main/src/library/Set.flix

```diff
@@ -1,7 +1,170 @@
+ * limitations under the License.
+ */
+
 namespace Set {
 
-    def empty[t]: Set[t] = Set(Nil)
+    ///
+    /// Returns the size of `xs`.
+    ///
+    def size[a](xs: Set[a]): Int32 =
+        let Set(s) = xs;
+            List/length(s)
+
+    //
+    // ## Set Construction
+    //
+
+    ///
+    /// Returns the empty set.
+    ///
+    def empty[a]: Set[a] = Set(Nil)
+
+    ///
+    /// Returns the singleton set containing `x`.
+    ///
+    def singleton[a](x: a): Set[a] = Set(x :: Nil)
+
+    ///
+    /// Adds `x` to `xs`.
+    ///
+    def insert[a](x: a, xs: Set[a]): Set[a] =
+        let Set(s) = xs;
+            if (List/memberOf(x, s)) xs else Set(x :: s)
+
+    ///
+    /// Removes `x` from `xs`.
+    ///
+    def delete[a](x: a, xs: Set[a]): Set[a] =
+        let Set(s) = xs;
+            Set(List/remove(x, s))
+
+    //
+    // ## Set Predicates
+    //
+
+    ///
+    /// Returns true if and only if `xs` is the empty set, i.e. `Set(Nil)`.
+    ///
+    def null[a](xs: Set[a]): Bool =
+        let Set(s) = xs;
+            List/isEmpty(s)
+
+    ///
+    /// Returns true if and only if `x` is a member of `xs`.
+    ///
+    def memberOf[a](x: a, xs: Set[a]): Bool =
+        let Set(s) = xs;
+            List/memberOf(x, s)
+
+    ///
+    /// Returns true if and only if every element in `xs` appears in `ys`.
+    ///
+    def isSubsetOf[a](xs: Set[a], ys: Set[a]): Bool = match xs with {
+        case Set(Nil) => true
+        case Set(x :: rs) => if (memberOf(x, ys)) isSubsetOf(Set(rs), ys) else false
+    }
+
+    ///
+    /// Returns true if and only if every element in `xs` appears in `ys` and `xs != ys`.
+    ///
+    def isProperSubsetOf[a](xs: Set[a], ys: Set[a]): Bool = if (isSubsetOf(xs, ys)) size(xs) != size(ys) else false
+
+    //
+    // ## Elementary Set Operations
+    //
+
+    ///
+    /// Returns the union of `xs` and `ys`.
+    ///
+    def union[a](xs: Set[a], ys: Set[a]): Set[a] = match xs with {
+        case Set(Nil) => ys
+        case Set(x :: rs) =>
+            let s = union(Set(rs), ys);
+                if (memberOf(x, ys)) s else insert(x, s)
+    }
+
+    ///
+    /// Returns the intersection of `xs` and `ys`.
+    ///
+    def intersection[a](xs: Set[a], ys: Set[a]): Set[a] = match xs with {
+        case Set(Nil) => xs
+        case Set(x :: rs) =>
+            let s = intersection(Set(rs), ys);
+                if (memberOf(x, ys)) insert(x, s) else s
+    }
+
+    ///
+    /// Returns the difference of `xs` and `ys`, i.e. `xs - ys`.
+    ///
+    def difference[a](xs: Set[a], ys: Set[a]): Set[a] = match xs with {
+        case Set(Nil) => xs
+        case Set(x :: rs) =>
+            let s = difference(Set(rs), ys);
+                if (memberOf(x, ys)) s else insert(x, s)
+    }
+
+    ///
+    /// Returns all subsets of `xs`.
+    ///
+    def subsets[a](xs: Set[a]): Set[Set[a]] =
+        let Set(s) = xs;
+        let f = x -> Set(x);
+            Set(List/map(f, List/subsequences(s)))
```

Comment:
Since a list could contain duplicates (and hence subsequences could return a list with duplicates) you should add a call to `toSet` somewhere here which takes care of duplicates.